File: erasers_kachaka/erasers_kachaka_cartographer/src/auto_map_saver.cpp
#include "rclcpp/rclcpp.hpp"
#include <cstdlib>
#include <chrono>
#include <thread>
#include <string>

class AutoMapSaver : public rclcpp::Node {
public:
    AutoMapSaver() : Node("auto_map_saver") {
        this->declare_parameter<std::string>("map_name", "map");
        this->declare_parameter<std::string>("map_save_path", "/home/roboworks/education_ws/src/erasers_kachaka/erasers_kachaka/erasers_kachaka_cartographer/map");
        this->declare_parameter<int>("save_late", 5);

        this->get_parameter("map_save_path", map_save_path_);
        this->get_parameter("map_name", map_name_);
        this->get_parameter("save_late", save_late_);

        namespace_ = this->get_namespace();

        timer_ = this->create_wall_timer(
            std::chrono::seconds(save_late_),
            std::bind(&AutoMapSaver::saveMap, this)
        );
    }

private:
    void saveMap() {
        std::string cmd = "ros2 run nav2_map_server map_saver_cli -f \"" + map_save_path_ + "/" + map_name_ + "\" --ros-args -p map_subscribe_transient_local:=true -r __ns:=" + namespace_;
        
        int result = std::system(cmd.c_str());
        if (result == 0) {
            RCLCPP_INFO(this->get_logger(), "Saved %s", map_save_path_.c_str());
        } else {
            RCLCPP_ERROR(this->get_logger(), "Map does not exist!");
        }
    }

    std::string map_save_path_;
    std::string map_name_;
    int save_late_;
    std::string namespace_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<AutoMapSaver>());
    rclcpp::shutdown();
    return 0;
}

File: erasers_kachaka/erasers_kachaka_cartographer/auto_map_saver.py
#!/usr/bin/env python3
from rclpy.node import Node
import rclpy

import subprocess
import time

def main():
    rclpy.init()
    node = Node("auto_map_saver")

    namespace = node.get_namespace()

    node.declare_parameter('map_name', 'map')
    node.declare_parameter('map_save_path', '/home/roboworks/education_ws/src/erasers_kachaka/erasers_kachaka/erasers_kachaka_cartographer/map')
    node.declare_parameter('save_late', 5)

    path = node.get_parameter('map_save_path').get_parameter_value().string_value
    name = node.get_parameter('map_name').get_parameter_value().string_value
    late = node.get_parameter('save_late').get_parameter_value().integer_value

    cmd1 = 'ros2 run nav2_map_server map_saver_cli -f \"%s/%s\" --ros-args -p map_subscribe_transient_local:=true -r __ns:=%s'%(path, name, namespace)

    while rclpy.ok():
        try:
            try:
                subprocess.run(cmd1,
                                shell=True,
                                check=True,
                                stdout=subprocess.PIPE,
                                stderr=subprocess.PIPE)
                node.get_logger().info('saved %s'%path)
            except subprocess.CalledProcessError as e:
                node.get_logger().error('map is not exists !')
            time.sleep(late)
            
        except KeyboardInterrupt:
            subprocess.run(cmd1,
                            shell=True,
                            stdout=subprocess.PIPE)
            node.get_logger().info('saved %s'%path)
            node.get_logger().info('Done ...')
            break

if __name__ == "__main__":
    main()